src/kernel-install/60-ukify.install.in

```diff
@@ -0,0 +1,203 @@
+        log(f'{KERNEL_INSTALL_LAYOUT=}, quitting.')
+        return False
+
+    KERNEL_INSTALL_UKI_GENERATOR = os.getenv('KERNEL_INSTALL_UKI_GENERATOR')
+
+    if KERNEL_INSTALL_UKI_GENERATOR != 'ukify':
+        log(f'{KERNEL_INSTALL_UKI_GENERATOR=}, quitting.')
+        return False
+
+    log('KERNEL_INSTALL_LAYOUT and KERNEL_INSTALL_UKI_GENERATOR are good')
+    return True
+
+def kernel_cmdline() -> str:
+    entry_token = os.environ['KERNEL_INSTALL_ENTRY_TOKEN']
+    machine_id = os.environ['KERNEL_INSTALL_MACHINE_ID']
+
+    if root := os.getenv('KERNEL_INSTALL_CONF_ROOT'):
+        try:
+            options = Path(root).joinpath('cmdline').read_text().split()
+        except FileNotFoundError:
+            return None
+    else:
+        for cmdline, fallback in [
+                ('/etc/kernel/cmdline',     False),
+                ('/usr/lib/kernel/cmdline', False),
+                ('/proc/cmdline',           True),
+        ]:
+            try:
+                options = Path(cmdline).read_text().split()
+            except FileNotFoundError:
+                if fallback:
+                    raise
+
+        if fallback:
+            options = [opt for opt in options
+                       if not opt.startswith(('BOOT_IMAGE=', 'initrd='))]
+
+    # If the boot entries are named after the machine ID, then suffix the kernel
+    # command line with the machine ID we use, so that the machine ID remains
+    # stable, even during factory reset, in the initrd (where the system's machine
+    # ID is not directly accessible yet), and if the root file system is volatile.
+    if (entry_token == machine_id and
+        not any(opt.startswith('systemd.machine_id=') for opt in options)):
+        options += [f'systemd.machine_id={machine_id}']
+
+    cmdline = ' '.join(options)
+    if cmdline.startswith('@'):
+        raise ValueError("Kernel options must not start with '@'")
+    return cmdline
+
+def make_uki(opts, cmdline):
+    output = opts.staging_area / 'uki.efi'
+
+    # Args must be strings here to not confuse argparse in ukify.parse_args().
+    args = [
+        opts.kernel_image,
+        *opts.initrd,
+        f'--uname={opts.kernel_version}',
+        f'--output={output}',
+        f'--cmdline={cmdline}',
+    ]
+
+    if BOOT_STUB:
+        args += [f'--stub={BOOT_STUB}']
+
+    # Punish me harder.
+    # We want this:
+    #   ukify = importlib.machinery.SourceFileLoader('ukify', UKIFY).load_module()
+    # but it throws a DeprecationWarning.
+    # https://stackoverflow.com/questions/67631/how-can-i-import-a-module-dynamically-given-the-full-path
+    # https://github.com/python/cpython/issues/65635
+    # offer "explanations", but to actually load a python file without a .py extension,
+    # the "solution" is 4+ incomprehensible lines.
+    # The solution with runpy gives a dictionary, which isn't great, but will do.
+    ukify = runpy.run_path(UKIFY, run_name='ukify')
+    ukify['main'](args)
```

Comment:
Don't be so hard on yourself, I actually think using `runpy` is a good and readable solution! :)

I do think though, that maybe this comment doesn't need to be shipped where everybody can read it. While I agree, that the one-liner is preferable and something like it should be added back to the stdlib, the alternative they provide as a recipe is not that inscrutable. The reason is that the Python devs wanted to split getting information about a module from loading it, which is why they added the module specs so now initialising a module is this three step process of getting the spec, making the module from the spec and then executing it in a clean namespace. Hooking the module up to `sys.modules` would probably not be necessary for ukify, since that seems to be needed to make relative imports work, but ukify is a single file.

That being said, I think I would have gone with your first solution of shelling out to `ukify`, as I don't really see an advantage in this dance, no matter which version.

Reply:
The first comment on the stackoverflow question has ~~138~~139 upvotes, and it expresses my sentiment very well. I don't understand why Python developers are trying so hard to make this hard and not listening to people telling them that the new "solution" is not great.

Not shelling out has advantages. For the next version of this PR I reworked the ukify code to skip the option-parsing step. This avoids pitfalls with syntax and is generally nicer. It's certainly more work, but I hope it'll be worth it in the long run.